node/orchestra/proc-macro/src/graph.rs

```diff
@@ -0,0 +1,423 @@
+		// create a directed graph with all the subsystems as nodes and the messages as edges
+		// key is always the message path, values are node indices in the graph and the subsystem generic identifier
+		// store the message path and the source sender, both in the graph as well as identifier
+		let mut outgoing_lut = HashMap::<&Path, Vec<(&Ident, NodeIndex)>>::with_capacity(128);
+		// same for consuming the incoming messages
+		let mut consuming_lut = HashMap::<&Path, (&Ident, NodeIndex)>::with_capacity(128);
+
+		let mut graph = Graph::<Ident, Path>::new();
+
+		// prepare the full index of outgoing and source subsystems
+		for ssf in ssfs {
+			let node_index = graph.add_node(ssf.generic.clone());
+			for outgoing in ssf.messages_to_send.iter() {
+				outgoing_lut.entry(outgoing).or_default().push((&ssf.generic, node_index));
+			}
+			if let Some(_first_consument) =
+				consuming_lut.insert(&ssf.message_to_consume, (&ssf.generic, node_index))
+			{
+				// bail, two subsystems consuming the same message
+			}
+		}
+
+		for (message_ty, (_consuming_subsystem_ident, consuming_node_index)) in consuming_lut.iter()
+		{
+			// match the outgoing ones that were registered above with the consumed message
+			if let Some(origin_subsystems) = outgoing_lut.get(message_ty) {
+				for (_origin_subsystem_ident, sending_node_index) in origin_subsystems.iter() {
+					graph.add_edge(
+						*sending_node_index,
+						*consuming_node_index,
+						(*message_ty).clone(),
+					);
+				}
+			}
+		}
+
+		// extract unsent and unreceived messages
+		let outgoing_set = HashSet::<_, RandomState>::from_iter(outgoing_lut.keys().cloned());
+		let consuming_set = HashSet::<_, RandomState>::from_iter(consuming_lut.keys().cloned());
+
+		let mut unsent_messages = consuming_lut;
+		unsent_messages.retain(|k, _v| !outgoing_set.contains(k));
+
+		let mut unconsumed_messages = outgoing_lut;
+		unconsumed_messages.retain(|k, _v| !consuming_set.contains(k));
+
+		let scc = Self::extract_scc(&graph);
+
+		Self { graph, sccs: scc, unsent_messages, unconsumed_messages }
+	}
+
+	/// Extract the strongly connected clusters (`scc`) which each
+	/// includes at least one cycle each.
+	fn extract_scc(graph: &Graph<Ident, Path>) -> Vec<Vec<NodeIndex>> {
+		use petgraph::visit::EdgeRef;
+
+		// there is no guarantee regarding the node indices in the individual sccs
+		let sccs = petgraph::algo::kosaraju_scc(&graph);
+		let sccs = Vec::from_iter(sccs.into_iter().filter(|scc| {
+			match scc.len() {
+				1 => {
+					// contains sccs of length one,
+					// which do not exists, might be an upstream bug?
+					let node_idx = scc[0];
+					graph
+						.edges_directed(node_idx, petgraph::Direction::Outgoing)
+						.find(|edge| edge.target() == node_idx)
+						.is_some()
+				},
+				0 => false,
+				_n => true,
+			}
+		}));
+		match sccs.len() {
+			0 => println!("✅ Found no strongly connected clusters, hence no cycles exist"),
```

Comment:
nit: a mixture in terminology "strongly connected component" vs "strongly connected cluster".
I haven't heard the latter, so the former is preferred: https://en.wikipedia.org/wiki/Strongly_connected_component.

Reply:
Good catch :+1: